src/opts.c

```diff
@@ -883,17 +883,190 @@ natsOptions_SetMaxPendingMsgs(natsOptions *opts, int maxPending)
+{
+    nats_CallbackList *this, *prev;
+
+    if (f == NULL)
+        return;
+
+    for (this = *headptr, prev = NULL; this != NULL; prev = this, this = this->next)
+    {
+        if (((void(*)(void))(this->f.conn) == f) && (this->closure == closure))
+        {
+            *prior_cb_removed = this;
+            if (prev == NULL) 
+            {
+                *headptr = this->next;
+            }
+            else
+            {
+                prev->next = this->next;
+            }
+            
+            return;
+        }
+    }
+}
+
+void
+natsOptions_freeCallbackList(nats_CallbackList *cb)
+{
+    nats_CallbackList *next = NULL;
+
+    while (cb != NULL)
+    {
+        next = cb->next;
+        NATS_FREE(cb);
+        cb = next;
+    }
+}
+
+natsStatus
+natsOptions_cloneCallbackList(nats_CallbackList **clone, nats_CallbackList *this)
+{
+    nats_CallbackList *prev = NULL;
+    nats_CallbackList *head = NULL;
+    nats_CallbackList *tail = NULL;
+
+    for (; this != NULL; this = this->next)
+    {
+        tail = NATS_CALLOC(1, sizeof(nats_CallbackList));
+        if (tail == NULL)
+        {
+            natsOptions_freeCallbackList(head);
+            return nats_setDefaultError(NATS_NO_MEMORY);
+        }
+        *tail = *this;
+        tail->next = NULL;
+
+        if (prev == NULL)
+        {
+            head = tail;
+        }
+        else
+        {
+            prev->next = tail;
+        }
+        prev = tail;
+    }
+
+    *clone = head;
+    return NATS_OK;
+}
+
+natsStatus
+natsOptions_addConnectionClosedCallback(natsOptions *opts, natsConnectionHandler f, void *closure)
+{
+    natsStatus          s           = NATS_OK;
+    nats_CallbackList   *cb         = NULL;
+    nats_CallbackList   *replaced   = NULL;
+
+    cb = NATS_CALLOC(1, sizeof(nats_CallbackList));
+    if (cb == NULL)
+        return nats_setDefaultError(NATS_NO_MEMORY);
+    cb->type = CALLBACK_TYPE_CONN;
+    cb->f.conn = f;
+    cb->closure = closure;
+
+    LOCK_AND_CHECK_OPTIONS(opts, 0);
+
+    // unlink if already in the list.
+    natsOptions_unlinkCallback(&replaced, &opts->closedCb, (void(*)(void))f, closure);
+
+    // add at the head.
+    cb->next = opts->closedCb;
+    opts->closedCb = cb;
+
+    UNLOCK_OPTS(opts);
+
+    NATS_FREE(replaced);
+
+    return s;
+}
+
+natsStatus
+natsOptions_addErrorCallback(natsOptions *opts, natsErrHandler f, void *closure)
+{
+    natsStatus          s           = NATS_OK;
+    nats_CallbackList   *cb         = NULL;
+    nats_CallbackList   *replaced   = NULL;
+
+    cb = NATS_CALLOC(1, sizeof(nats_CallbackList));
+    if (cb == NULL)
+        return nats_setDefaultError(NATS_NO_MEMORY);
+    cb->type = CALLBACK_TYPE_ERROR;
+    cb->f.err = f;
+    cb->closure = closure;
+
+    LOCK_AND_CHECK_OPTIONS(opts, 0);
+
+    // unlink if already in the list.
+    natsOptions_unlinkCallback(&replaced, &opts->asyncErrCb, (void(*)(void))f, closure);
```

Comment:
You are missing a free for `replaced`

Reply:
oops, thx! fixed. 0/5 if it's worth the time adding tests for these new methods. Perhaps if they get another use-case and/or get publicly exposed?